helpers.ts

```diff
@@ -337,6 +338,16 @@ export function isPromise(candidateFuture: any): boolean {
 	return !!(candidateFuture && typeof (candidateFuture.then) === "function");
 }
 
+export async function attachAwaitDetach(eventName: string, ee: EventEmitter, cb: Function, operation: Promise<any>) {
+	ee.on(eventName, cb);
```

Comment:
I think this might be a bit of an overkill for this particular task. I really do like the idea though, in fact I've considered something similar in the past but never got around to implementing it.

A concern of mine is the fact that we already have `IDisposable` interface and a mechanism in our dependency injection for calling `.dispose()`. However you certainly wouldn't want to call `.dispose()` on the classes that currently implement `IDisposable` yourself. That would create a bit of confusion IMO.